processor/src/trace/virtual_bus/multilinear/mod.rs

```diff
@@ -0,0 +1,151 @@
+        Ok(Self {
+            num_variables: log2(values.len()) as usize,
+            evaluations: values,
+        })
+    }
+
+    /// Constructs a [MultiLinear] from a slice of values in the field.
+    pub fn from_values(values: &[E]) -> Result<Self, Error> {
+        if !values.len().is_power_of_two() {
+            return Err(Error::EvaluationsNotPowerOfTwo);
+        }
+        Ok(Self {
+            num_variables: log2(values.len()) as usize,
+            evaluations: values.to_owned(),
+        })
+    }
+
+    /// Returns the number of the multi-linear polynomial.
+    pub fn num_variables(&self) -> usize {
+        self.num_variables
+    }
+
+    /// Returns the evaluations over the boolean hyper-cube.
+    pub fn evaluations(&self) -> &[E] {
+        &self.evaluations
+    }
+
+    /// Returns the number of evaluations. This is equal to the size of the boolean hyper-cube.
+    pub fn num_evaluations(&self) -> usize {
+        self.evaluations.len()
+    }
+
+    /// Evaluate the multi-linear at some query (r_0, ..., r_{ν - 1}) ∈ 𝔽^ν.
+    ///
+    /// It first computes the evaluations of the Lagrange basis polynomials over the interpolating
+    /// set {0 , 1}^ν at (r_0, ..., r_{ν - 1}) i.e., the Lagrange kernel at (r_0, ..., r_{ν - 1}).
+    /// The evaluation then is the inner product, indexed by {0 , 1}^ν, of the vector of
+    /// evaluations times the Lagrange kernel.
+    pub fn evaluate(&self, query: &[E]) -> E {
+        let tensored_query = tensorize(query);
+        inner_product(&self.evaluations, &tensored_query)
+    }
+
+    /// Computes f(r_0, y_1, ..., y_{ν - 1}) using the linear interpolation formula
+    /// (1 - r_0) * f(0, y_1, ..., y_{ν - 1}) + r_0 * f(1, y_1, ..., y_{ν - 1}). The resulting
+    /// returned multi-linear is defined over a domain of half the size.
+    pub fn bind(&self, round_challenge: E) -> Self {
+        let mut result = vec![E::ZERO; 1 << (self.num_variables() - 1)];
+        for (i, res) in result.iter_mut().enumerate() {
+            *res = self.evaluations[i << 1]
+                + round_challenge * (self.evaluations[(i << 1) + 1] - self.evaluations[i << 1]);
+        }
+        Self::from_values(&result).expect("should not fail given that it is a multi-linear")
+    }
+
+    /// Computes f(r_0, y_1, ..., y_{ν - 1}) using the linear interpolation formula
+    /// (1 - r_0) * f(0, y_1, ..., y_{ν - 1}) + r_0 * f(1, y_1, ..., y_{ν - 1}) and assigns
+    /// the resulting multi-linear, defined over a domain of half the size, to `self`.
+    pub fn bind_assign(&mut self, round_challenge: E) {
+        let mut result = vec![E::ZERO; 1 << (self.num_variables() - 1)];
+        for (i, res) in result.iter_mut().enumerate() {
+            *res = self.evaluations[i << 1]
+                + round_challenge * (self.evaluations[(i << 1) + 1] - self.evaluations[i << 1]);
+        }
+        *self =
+            Self::from_values(&result).expect("should not fail given that it is a multi-linear");
```

Comment:
`bind_assign(&mut self)` to me suggests that this is done in-place, and that there is a performance benefit over using `bind()`. Instead, this is a new function for `*self = self.bind(...)`. I would remove this function and use `bind()` everywhere instead

Reply:
I would like it to be in place but not sure about the most straightforward way to do this
I removed and renamed as suggested